### PR TITLE
Dynamic execution duration for leap

### PR DESCRIPTION
**Pull request by AminArria**
Changes leap from having a fixed duration and dynamic speed to having fixed speed and dynamic duration

### CODE REVIEW

File: apps/arena/lib/arena/game/player.ex [hunk 1]
```diff
@@ -362,4 +357,23 @@ defmodule Arena.Game.Player do
         player
     end
   end
+
+  ## Yes, we are pattern matching on exactly one mechanic. As of time writing we only have one mechanic per skill
+  ## so to simplify my life an executive decision was made to take thas as a fact
+  ## When the time comes to have more than one mechanic per skill this function will need to be refactored, good thing
+  ## is that it will crash here so not something we can ignore
+  defp calculate_duration(%{mechanics: [%{leap: leap}]}, position, direction) do
```
Comment (manucamejo) [inline apps/arena/lib/arena/game/player.ex:365]: This is not working. It should be:
`defp calculate_duration(%{mechanics: [leap: leap]}, position, direction) do`

Comment (manucamejo) [inline apps/arena/lib/arena/game/player.ex:365]: This is not working. It should be:
`defp calculate_duration(%{mechanics: [leap: leap]}, position, direction) do`

**Review by manucamejo — CHANGES_REQUESTED** [review]

File: apps/arena/lib/arena/game/player.ex (hunk 1 above)
Comment (AminArria) [reply apps/arena/lib/arena/game/player.ex]: Fixed https://github.com/lambdaclass/mirra_backend/pull/390/commits/7864206913ebb60788f65b02aedf1a7e47d19e62

File: apps/arena/lib/arena/game/player.ex [hunk 2]
```diff
@@ -396,6 +391,25 @@ defmodule Arena.Game.Player do
...
+  ## is that it will crash here so not something we can ignore
+  defp calculate_duration(%{mechanics: [{:leap, leap}]}, position, direction) do
+    ## TODO: Cap target_position to leap.range
+    target_position = %{
+      x: position.x + direction.x * leap.range,
+      y: position.y + direction.y * leap.range
+    }
+
+    ## TODO: Magic number needs to be replaced with state.game_config.game.tick_rate_ms
+    Physics.calculate_duration(position, target_position, leap.speed) * 30
```
Comment (agustinesco) [inline apps/arena/lib/arena/game/player.ex:406]: not in this pr but maybe we can aslo include this number in the config.exs file

File: apps/arena/lib/arena/game/player.ex [hunk 3]
```diff
@@ -396,6 +391,25 @@ defmodule Arena.Game.Player do
     end
   end
 
+  ## Yes, we are pattern matching on exactly one mechanic. As of time writing we only have one mechanic per skill
+  ## so to simplify my life an executive decision was made to take thas as a fact
+  ## When the time comes to have more than one mechanic per skill this function will need to be refactored, good thing
+  ## is that it will crash here so not something we can ignore
+  defp calculate_duration(%{mechanics: [{:leap, leap}]}, position, direction) do
+    ## TODO: Cap target_position to leap.range
```
Comment (agustinesco) [inline apps/arena/lib/arena/game/player.ex:399]: For what i know the distance is already caped at the leap range right?

File: apps/arena/lib/arena/game/skill.ex [hunk 4]
```diff
@@ -268,14 +262,21 @@ defmodule Arena.Game.Skill do
...
 
     ## TODO: Magic number needs to be replaced with state.game_config.game.tick_rate_ms
-    speed = Physics.calculate_speed(entity.position, target_position, leap.duration_ms) * 30
+    duration = Physics.calculate_duration(entity.position, target_position, leap.speed) * 30
+
+    Process.send_after(
+      self(),
+      {:stop_leap, entity.id, entity.aditional_info.base_speed, leap.on_arrival_mechanic},
+      duration
+    )
```
Comment (agustinesco) [inline apps/arena/lib/arena/game/skill.ex:271]: we're calculating the duration twice, maybe we could include it in the skill params map that `do_mechanic` receives and use that, what do you think? To do so you need to add the duration to the send message of `delayed_skill_mechanics`

File: apps/arena/priv/config.json [hunk 5]
```diff
@@ -948,8 +948,8 @@
       "mechanics": [
         {
           "leap": {
-            "range": 1500.0,
-            "duration_ms": 600,
+            "range": 1000.0,
```
Comment (agustinesco) [inline apps/arena/priv/config.json:951]: Why reducing the range?

File: apps/arena/lib/arena/game/player.ex (hunk 3 above)
Comment (AminArria) [reply apps/arena/lib/arena/game/player.ex:399]: Nope, there is no backend validation on the actual position you are sending. Meaning that a hacked client could in theory send anything

File: apps/arena/lib/arena/game/skill.ex (hunk 4 above)
Comment (AminArria) [reply apps/arena/lib/arena/game/skill.ex]: Nice catch, moved it as you said

File: apps/arena/priv/config.json (hunk 5 above)
Comment (AminArria) [reply apps/arena/priv/config.json]: Because merging is hard... Fixed now

Comment (agustinesco) [reply apps/arena/priv/config.json]: 😂

**Review by agustinesco — CHANGES_REQUESTED** [review]
I've found another error i think that, if you leap with the magic boots the final positions ends up being farther that you intended

https://github.com/lambdaclass/mirra_backend/assets/106101218/654d3e6b-e68f-49e4-a71a-27f41ad555b4

**Review by agustinesco — APPROVED** [review]
LGTM!

**Review by manucamejo — APPROVED** [review]

**Review by agustinesco — APPROVED** [review]

**Pull request merged**